### PR TITLE
run: add new --pipe option for "systemd-run"

### PR DESCRIPTION
**Pull request by poettering**
In this mode, we'll directly connect stdin/stdout/stderr of the invoked
service with whatever systemd-run itself is invoked on. This allows
inclusion of "systemd-run" commands in shell pipelines, as unlike
"--pty" this means EOF of stdin/stdout/stderr are propagated
independently.

If --pty and --pipe are combined systemd-run will automatically pick the
right choice for the context it is invoked in, i.e. --pty when invoked
on a TTY, and --pipe otherwise.

### CODE REVIEW

File: man/systemd-run.xml [hunk 1]
```diff
@@ -221,12 +221,31 @@
 
         <listitem><para>When invoking the command, the transient service connects its standard input and output to the
         terminal <command>systemd-run</command> is invoked on, via a pseudo TTY device. This allows running binaries
-        that expect interactive user input as services, such as interactive command shells.</para>
+        that expect interactive user input/input as services, such as interactive command shells.</para>
```
Comment (keszybz) [inline man/systemd-run.xml]: What does this slash mean? It's unclear.

File: man/systemd-run.xml [hunk 2]
```diff
@@ -221,12 +221,31 @@
...
+
+        <para>See below for details on how this switch combines with <option>--pipe</option>.</para></listitem>
+      </varlistentry>
+
+      <varlistentry>
+        <term><option>--pipe</option></term>
+        <term><option>-P</option></term>
+
+        <listitem><para>When invoking the command, the transient service's standard input, standard output and standard
+        error are directly connected to the standard inpout, standard output and standard error the
```
Comment (keszybz) [inline man/systemd-run.xml]: Typo in "inpout".

Maybe it's easier to say "Standard input, output, and error of the transient service are inherited from the `systemd-run` command itself.". That's what is e.g. used in Python subprocess.Popen docs, and it should be pretty clear for most readers.

File: src/run/run.c [hunk 3]
```diff
@@ -545,6 +572,24 @@ static int transient_service_set_properties(sd_bus_message *m, char **argv, cons
...
+
+                r = sd_bus_message_append(m,
+                                          "(sv)(sv)(sv)",
+                                          "StandardInputFileDescriptor", "h", STDIN_FILENO,
+                                          "StandardOutputFileDescriptor", "h", STDOUT_FILENO,
+                                          "StandardErrorFileDescriptor", "h", STDERR_FILENO);
+                if (r < 0)
+                        return r;
+
+
```
Comment (keszybz) [inline src/run/run.c]: Spurious newline.

File: src/run/run.c [hunk 4]
```diff
@@ -373,6 +391,16 @@ static int parse_argv(int argc, char *argv[]) {
...
                 }
 
+
+        if (arg_stdio == ARG_STDIO_AUTO) {
+                /* If we both --pty and --pipe are specified we'll automatically pick --pty if we are connected fully
+                 * to a TTY and pick direct fd passing otherwise. This way, we automatically adapt to usage in a shell
+                 * pipeline, but we are neatly interactive with tty-level isolation otherwise. */
+                arg_stdio = isatty(STDIN_FILENO) && isatty(STDOUT_FILENO) && isatty(STDERR_FILENO) ?
+                        ARG_STDIO_PTY :
+                        ARG_STDIO_DIRECT;
```
Comment (keszybz) [inline src/run/run.c:401]: That can be a bit tricky, bit it *should* work. As long as it's documented, users should not be surprised.

File: man/systemd-run.xml (hunk 1 above)
Comment (poettering) [reply man/systemd-run.xml]: oops, that's a typo. was supposed to say "input/output", as in "I/O"

**Comment by poettering** [comment]
Force pushed a new versions, with all issues @keszybz pointed out addressed. Please have a another look!

**Comment by poettering** [comment]
CI failure appears unrelated

File: man/systemd-run.xml [hunk 5]
```diff
@@ -219,14 +219,31 @@
         <term><option>--pty</option></term>
         <term><option>-t</option></term>
 
-        <listitem><para>When invoking the command, the transient service connects its standard input and output to the
-        terminal <command>systemd-run</command> is invoked on, via a pseudo TTY device. This allows running binaries
-        that expect interactive user input as services, such as interactive command shells.</para>
+        <listitem><para>When invoking the command, the transient service connects its standard input, output and error
+        to the terminal <command>systemd-run</command> is invoked on, via a pseudo TTY device. This allows running
+        binaries that expect interactive user input/output as services, such as interactive command shells.</para>
```
Comment (keszybz) [inline man/systemd-run.xml]: s/binaries/programs/. E.g. python scripts are OK too.

File: man/systemd-run.xml [hunk 6]
```diff
@@ -219,14 +219,31 @@
...
+
+        <para>See below for details on how this switch combines with <option>--pipe</option>.</para></listitem>
+      </varlistentry>
+
+      <varlistentry>
+        <term><option>--pipe</option></term>
+        <term><option>-P</option></term>
+
+        <listitem><para>If specified, standard input, output, and error of the transient service are inherited from the
+        <command>systemd-run</command> command itself. Use this option in order to use <command>systemd-run</command>
```
Comment (keszybz) [inline man/systemd-run.xml]: "Use ... use", awkard.

**Pull request merged**